src/HIE/Bios/Cradle.hs

```diff
@@ -435,12 +437,107 @@ cabalCradle wdir mc =
+      pure $ processEnvironment ghcDirs ++ environment
+
+    setupCabalCommand :: (FilePath, FilePath) -> IO CreateProcess
+    setupCabalCommand (ghcBin, libdir) = do
+      wrapper_fp <- withGhcWrapperTool ("ghc", []) workDir
+      buildDir <- cabalBuildDir workDir
+      ghcPkgPath <- withGhcPkgTool ghcBin libdir
+      let extraCabalArgs =
+            [ "--builddir=" <> buildDir
+            , command
+            , "--with-compiler", wrapper_fp
+            , "--with-hc-pkg", ghcPkgPath
+            ]
+      pure $ proc "cabal" (extraCabalArgs ++ args)
+
+-- | Discovers the location of 'ghc-pkg' given the absolute path to 'ghc'
+-- and its '$libdir' (obtainable by running @ghc --print-libdir@).
+--
+-- @'withGhcPkgTool' ghcPathAbs libdir@ guesses the location by looking at
+-- the filename of 'ghcPathAbs' and expects that 'ghc-pkg' is right next to it,
+-- which is guaranteed by the ghc build system. Most OS's follow this
+-- convention.
+--
+-- On unix, there is a high-chance that the obtained 'ghc' location is the
+-- "unwrapped" executable, e.g. the executable without a shim that specifies
+-- the '$libdir' and other important constants.
+-- As such, the executable 'ghc-pkg' is similarly without a wrapper shim and
+-- is lacking certain constants such as 'global-package-db'. It is, therefore,
+-- not suitable to pass in to other consumers, such as 'cabal'.
+--
+-- Here, we restore the wrapper-shims, if necessary, thus the returned filepath
+-- can be passed to 'cabal' without further modifications.
+withGhcPkgTool :: FilePath -> FilePath -> IO FilePath
+withGhcPkgTool ghcPathAbs libdir = do
+  let ghcName = takeFileName ghcPathAbs
+      -- TODO: check for existence
+      ghcPkgPath = guessGhcPkgFromGhc ghcName
+  if isWindows
+    then pure ghcPkgPath
+    else withWrapperTool ghcPkgPath
+  where
+    ghcDir = takeDirectory ghcPathAbs
+
+    guessGhcPkgFromGhc ghcName =
+      let ghcPkgName = T.replace "ghc" "ghc-pkg" (T.pack ghcName)
+      in ghcDir </> T.unpack ghcPkgName
+
+    -- Only on unix, creates a wrapper script that's hopefully identical
+    -- to the wrapper script 'ghc-pkg' usually comes with.
+    --
+    -- 'ghc-pkg' needs to know the 'global-package-db' location which is
+    -- passed in via a wrapper shim that basically wraps 'ghc-pkg' and
+    -- only passes in the correct 'global-package-db'.
+    -- For an example on how the wrapper script is supposed to look like, take
+    -- a look at @cat $(which ghc-pkg)@, assuming 'ghc-pkg' is on your $PATH.
+    --
+    -- If we used the raw executable, i.e. not wrapped in a shim, then 'cabal'
+    -- can not use the given 'ghc-pkg'.
+    withWrapperTool ghcPkg = do
+      let globalPackageDb = libdir </> "package.conf.d"
+          -- This is the same as the wrapper-shims ghc-pkg usually comes with.
+          contents = unlines
+            [ "#!/bin/sh"
+            , unwords ["exec", ghcPkg, "--global-package-db", globalPackageDb, "${1+\"$@\"}"]
```

Comment:
```suggestion
            , unwords ["exec", "\"" <> ghcPkg <> "\"", "--global-package-db", "\"" <> globalPackageDb <> "\"", "${1+\"$@\"}"]
```

Here we are missing quotes for both `ghcPkg` and `globalPackageDb`

The resulting wrapper script looks like this:

```sh
#!/bin/sh
exec /home/hasufell/.ghcup/ghc/8.8.4/lib/ghc-8.8.4/bin/ghc-pkg --global-package-db /home/hasufell/.ghcup/ghc/8.8.4/lib/ghc-8.8.4/package.conf.d ${1+"$@"}
```

Reply:
since unix filepaths can contain `"`, we should call `show` on our strings, right?

Reply:
Conisder [`GHC.ResponseFile.escapeArgs`](https://hackage.haskell.org/package/base-4.14.1.0/docs/GHC-ResponseFile.html#v:escapeArgs)

Reply:
escapeArgs adds newlines, which I have trimmed away. I think this also might trim away other potentially correct paths from a filepath, but I think it suffices to patch it *if* it breaks on someones system